Support STR() Transact-SQL function.

LareinaWei opened the pull request:
Task: BABAL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

### Description

Add support for STR() Transact-SQL function.
T-SQL STR() function supports up to 17 digits of significant number, and up to 16 digit precision. 
- When input float expression has more than 17 significant digits, STR() will keep the first 17 significant digits and pad the rest of the digits with zeros. 
- When input float expression has more than 16 digits after decimal point, STR() will keep the first 16 digits and add number of rest digits preceding spaces in the return string.


### Issues Resolved
Babelfish do not support STR() Transact-SQL fucntion.

### Reference Document Link
https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/t-sql/functions/str-transact-sql.md

### Test Scenarios Covered 
* **Use case based -**
  Added a new test `functions/str`
* **Boundary conditions -**
  Covered in test `functions/str`
* **Arbitrary inputs -**
 added
* **Negative test cases -**
  Int overflow would thrown an error.
* **Minor version upgrade tests -**
  tests added
* **Major version upgrade tests -**
  tests added
* **Performance tests -**
  N/A
* **Tooling impact -**
  None
* **Client tests -**
  None


### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support STR() Transact-SQL function.

Task: BABAL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Add str test to JDBC test schedule

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
empty commit

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
removed covered functions from expected_dependency.out

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
removed some unused import from string.c

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Modify bracket convension

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
Add test case for NaN and Infinity input

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
empty commit

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

LareinaWei commented on contrib/babelfishpg_common/src/instr.h:
```diff
@@ -59,6 +59,7 @@ typedef enum PgTsqlInstrMetricType {
 	INSTR_TSQL_CEILING,
 	INSTR_TSQL_FLOOR,
 	INSTR_TSQL_ROUND,
+    INSTR_TSQL_STR,
```
I'll fix the indentation here and in string.c in the next commit...

Commit:
Add some examples and explanations in the comments

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
empty commit

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+    
+    // max precision is 17, max significant digits in decimal part = min(remaining significant digits, input decimal digits)
+    deci_sig = (17 > int_digits ? 17 - int_digits : 0);
+    deci_sig = (deci_sig > input_deci_digits ? input_deci_digits : deci_sig);
```
Can we use `Min/Max` macros in `c.h`?

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+            // the space limitation goes by the one set before the rounding & carried over
+            // STR(9999.998, 7, 2) returns "*******" but STR(10000.000, 7, 2) returns "10000.0"
+            // which means the max length constraint of integer part is still 4 after rounding
+            buf = palloc(size);
```
Why do we need to re-alloc the buf?

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+            // STR(9999.998, 7, 2) returns "*******" but STR(10000.000, 7, 2) returns "10000.0"
+            // which means the max length constraint of integer part is still 4 after rounding
+            buf = palloc(size);
+	        memset(buf, '*', size);
```
indentation

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+        }
+    }
+    
+    result = (*common_utility_plugin_ptr->tsql_varchar_input)(buf, size, -1);
```
1. Is it okay not to have `buf[size - 1] = '\0';`?
2. The below code piece is called several times in this function. You may consider introducing a separate function.
```
buf[size - 1] = '\0';
result = (*common_utility_plugin_ptr->tsql_varchar_input)(buf, size, -1);
pfree(buf);
PG_RETURN_VARCHAR_P(result);
```

LareinaWei replied:
> Is it okay not to have buf[size - 1] = '\0';?

Yes, `tsql_varchar_input()`  function calls `varchar_input()` which can take string input wether or not it is null terminated.

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+        // exceeds the max precision, need to round to 17th digit(excluding - and .) 
+        if (int_digits > 17) 
+        {   // round in int part
+            // STR(12345678901234567890, 20) returns "1234567890123456800"
```
I got `12345678901234567000` for `select STR(12345678901234567890, 20);`. Could you double check this?

LareinaWei replied:
The t-sql STR() function gets a little flaky when it rounds in the integer parts...

`SELECT  STR(123456789012345670, 20);`  returns `  123456789012345660`, which makes no sense..
`SELECT  STR(123456789012345671, 20);` and `SELECT  STR(123456789012345672, 20);` returns the same result as above.

Starting from 123456789012345673 to 123456789012345679, 
`SELECT  STR(12345678901234567X, 20);`  returns `  123456789012345680`, the last digit get rounded up

When integer has 19 digits,
`SELECT  STR(1234567890123456100, 20);` returns ` 1234567890123456000`, 17th digit `1` got rounded down 
`SELECT  STR(1234567890123456200, 20);` returns ` 1234567890123456300`, 17th digit `2` got rounded up 
`SELECT  STR(1234567890123456380, 20);`  returns ` 1234567890123456300` but `SELECT  STR(1234567890123456390, 20);`  returns ` 1234567890123456500`...


When integer has 20 digits,
`SELECT  STR(12345678901234562000, 20);` returns `12345678901234561000`
`SELECT  STR(12345678901234562050, 20);` returns `12345678901234563000`
`SELECT  STR(12345678901234567890, 20);`  returns `12345678901234567000`
`SELECT  STR(12345678901234569999, 20);`  returns `12345678901234569000`
all weird behaviors.

I think this would be a little bit too hard to recreate so I just went with the correct rounding...

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+        if (int_digits > 17) 
+        {   // round in int part
+            // STR(12345678901234567890, 20) returns "1234567890123456800"
+            curr_digit = float_char[17 + has_neg_sign] - 48;
```
Could you elaborate where `48` came from?

LareinaWei replied:
48 is the ascii for char 0. I should probably also change these  to `'0'`.

2jungkook commented on contrib/babelfishpg_tsql/src/string.c:
```diff
@@ -495,3 +498,284 @@ tsql_varchar_substr(PG_FUNCTION_ARGS)
+    {
+        // do rounding 
+        carry = 1;
+        while (round_pos > (0 + has_neg_sign) && carry) 
```
I feel like this is re-inventing the wheels. Can we leverage existing `round()` function? For example, if `round_pos` is 2, then we may use the following:
```
1234 --> 12.34 --> round(12.34) --> 12.00 --> 1200
```

Commit:
Seperate some parts of STR() function into several static helper
functions, add explanation and some test case about rounding in
integer part.

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

2jungkook commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -366,5 +366,6 @@ TestVariableDataLength
 BABEL-3801
 app_name
 atn2
+str
```
Please add this to 14_7/schedule as well.

Commit:
add test str to 14_7/schedule

Task: BABEL-398
Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
empty commit

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>

Commit:
empty commit

Signed-off-by: Yuhao Wei <yuhaowei@amazon.com>
